mri/reconstructors/self_calibrating.py

```diff
@@ -0,0 +1,180 @@
+# -*- coding: utf-8 -*-
+##########################################################################
+# pySAP - Copyright (C) CEA, 2017 - 2018
+# Distributed under the terms of the CeCILL-B license, as published by
+# the CEA-CNRS-INRIA. Refer to the LICENSE file or to
+# http://www.cecill.info/licences/Licence_CeCILL-B_V1-en.html
+# for details.
+##########################################################################
+
+"""
+This implements the self-calibrating reconstruction for the multi-channel case.
+"""
+
+# Module import
+from .base import ReconstructorBase
+from ..operators import GradSelfCalibrationSynthesis, \
+    GradSelfCalibrationAnalysis, WaveletN
+from .utils.extract_sensitivity_maps import get_Smaps
+
+
+class SelfCalibrationReconstructor(ReconstructorBase):
+    """ Self Calibrating reconstruction for multi-channel acquisition.
+    The coil sensitivity is estimated from a small portion of the  k-space
+    center and used to reconstruct the complex image.
+
+    For the Analysis case finds the solution for x of:
+        (1/2) * sum(||F Sl x - yl||^2_2, n_coils) + mu * ||Wt x||_1
+
+    For the Synthesise case finds the solution of:
+        (1/2) * sum(||F Sl Wt alpha - yl||^2_2, n_coils) + mu * ||alpha||_1
+
+    The sensitivity information is taken to be the low-resolution of the image
+    extractes from the k-space portion given in the parameter
+
+    Parameters
+    ----------
+    fourier_op: object of class FFT, NonCartesianFFT or Stacked3DNFFT in
+                mri.operators
+        Defines the fourier operator F in the above equation.
+    linear_op: object, (optional, default None)
+        Defines the linear sparsifying operator Wt. This must operate on x and
+        have 2 functions, op(x) and adj_op(coeff) which implements the
+        operator and adjoint opertaor. For wavelets, this can be object of
+        class WaveletN or WaveletUD2 from mri.operators .
+        If None, sym8 wavelet with nb_scales=3 is chosen.
+    prox_op: operator, (optional default None)
+        Defines the proximity operator for the regularization function.
+        For example, for L1 Norm, the proximity operator is Thresholding
+        If None, the proximity opertaor is defined as Soft thresholding
+        of wavelet coefficients with mu value as specified.
+    mu: float or np.ndarray, (optional, default 0)
+        If prox_op is None, the value of mu is used to form a proximity
+        operator that is soft thresholding of the wavelet coefficients.
+        If prox_op is specified, this is ignored.
+    lips_calc_max_iter: int, default 10
```

Comment:
`gradient_method` is missing and also please consider changing the name to `gradient_formulation`